enhancements/rhcos/extensions.md

```diff
@@ -0,0 +1,96 @@
+---
+title: support-for-coreos-extensions
+authors:
+  - "@cgwalters"
+reviewers:
+  - "@ashcrow"
+approvers:
+  - "@ashcrow"
+  - "@crawford"
+  - "@imcleod"
+  - "@runcom"
+creation-date: 2020-04-21
+last-updated: 2020-04-21
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support For CoreOS extensions
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes a MachineConfig fragment like:
+```
+extensions:
+  - usbguard
+```
+
+This is the OpenShift version of the [Fedora CoreOS extension system tracker](https://github.com/coreos/fedora-coreos-tracker/issues/401).
+
+That will add additional software onto the host - but this software will still be versioned with the host (included as part of the OpenShift release payload) and upgraded with the cluster.
+
+## Motivation
+
+We have requests to ship things like `usbguard` that are needed to meet compliance in some scenarios (and `usbguard` is very useful on bare metal), but `usbguard` also makes no sense to ship in an OS for the public cloud.
+
+The `rpm-ostree` project has had as its goal from the start to re-cast RPMs as "operating system extensions" - much like how e.g. Firefox and its extensions work.  By default it operates as a pure image system, but packages can be layered on (and overridden) client side.
+
+OpenShift is already making use of this today for the [realtime kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md).
+
+We propose continuing the trail blazed with the RT kernel by adding additional RPMs to `machine-os-content` that aren't part of the base OS "image", but can be installed later.  This is how `kernel-rt` works today; we added a `kernel-rt.rpm` that is committed to the container.  In the future though, we may instead ship a separate `machine-os-extensions` container, or something else.  The only API stable piece here is the `MachineConfig` field (same as for the `kernelType: realtime`).
```

Comment:
If we're going this route, why not publish all OS content into a container as a repo, install the content from said repo.  Then we can support RHEL utilizing the same source of packages published per version.

Reply:
First: that's a *much* *much* larger set of stuff, including a lot of stuff that has no business ever being installed on RHCOS (e.g. `ruby`, `postgresql`, `gcc` and `gnome-shell`).  The RHEL 8 DVD is 6.8GB.

The scale of this proposal is probably (hopefully) at most 50-100MB of additional stuff, which is an order of magnitude or two smaller in size.  And by package count it's probably 20 packages tops versus thousands.

Second...people who have reason to access all of that should have entitlements set up anyways.

Reply:
In addition to usbguard and real-time kernel, there are use cases for this regarding kata (see https://github.com/harche/enhancements-1/blob/66675b4dd1fde442504ac331bc6343d90cfe1f1f/enhancements/kata/kata-operator.md) and, presumably, KubeVirt. In both cases, you need a `qemu` suitable for that version of RHCOS, and at least in the case of kata, a number of additional binaries.

Reply:
KubeVirt runs qemu in a container, it's not needed on the host

Reply:
@cgwalters I didn't mean package all of RHEL.  I mean all of OpenShift.  So whatever we're packing into CoreOS today, make those the same type of installation method as the 'extensions'.

Reply:
> @cgwalters I didn't mean package all of RHEL. I mean all of OpenShift. So whatever we're packing into CoreOS today, make those the same type of installation method as the 'extensions'.

Ah.  Well...the thing is I'd really like to keep the default RHCOS being in "image mode" by default where we're doing the depsolving, selinux labeling etc. all on the server and not per machine.

This of course gets into https://github.com/projectatomic/rpm-ostree/issues/1237

I see what you're thinking here but the counter is that if you want your OS content lifecycled with the cluster...that's what RHCOS is.

Reply:
> This of course gets into coreos/rpm-ostree#1237

This is an undecipherable rabbit hole for me.  I'm sure it makes a lot of sense to people heavily involved with these things, but I find it quite unapproachable.

> that's what RHCOS is

I fail to see a distinction.  We have some process that pulls in RPM content and delivers them to a host.  The only thing that is different is the underlying implementation.

If we add these 'extensions' we're opening ourselves up to the exact same problems RHEL has when it comes to shipping content.  Some package the user installs introduces a conflict.  Most packages also have some sort of configuration.  How are users going to handle configuration conflicts across versions?  Not to say these problems are blockers, but they seem to diminish the utility of RHCOS vs RHEL.

There's another proposal about building RPM ostrees in-cluster, and then also uploading images to the cloud for each release.  We might as well just build RHEL images at this point.  We might as well run a repo pod and stream just the bits that are updated rather than a whole ostree to each host.

I think the OStree model is good and has it's place.  Our software delivery model doesn't seem to necessarily be the best match, and it seems we're slowly rebuilding RHEL.

Reply:
Sorry I meant to link https://github.com/coreos/rpm-ostree/issues/1081

> Some package the user installs introduces a conflict. 

Conflict with what?  We own the set of extensions and testing them.

>  How are users going to handle configuration conflicts across versions?

Do you have a more specific exmaple?  Something like usbguard changing its configuration file incompatibly?  That seems like a "don't do that" situation.  But...I can imagine eventually we will need some sort of special handling in the MCO for this.

> There's another proposal about building RPM ostrees in-cluster, and then also uploading images to the cloud for each release. We might as well just build RHEL images at this point. We might as well run a repo pod and stream just the bits that are updated rather than a whole ostree to each host.

I won't attach my name to an operating system that leaves the root filesystem in a corrupted state if the kernel freezes or you run ENOSPC etc. during an update.  There are many other benefits to an "image based" system by default, among them clear versioning - typing `rpm-ostree status` shows you very clearly any deltas from the base image, you can always reset back to the base.

By default, you are running an *exact set of bits* that has been tested in CI rather than dynamically reassembled on each machine.

And notably with this extension proposal...we would probably start without any enabled extensions, so there's still no client-side depsolving.

Reply:
Generic lifecycle of RPMs on top of RHCoS should be completely out of scope.  We aren't shipping RHEL with OpenShift - we're shipping the package set that is needed to run containers, and everything else should be in a container.

Reply:
Are we just using this to avoid the real issue, of defining, what goes on the host and what goes into a container? 
as (hinted at by) https://github.com/coreos/fedora-coreos-tracker/issues/401#issue-571600267